File: algorithms/290.Word_Pattern/Word_Pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        # can be solved by two dictionaries
        a = list(pattern)
        b = str.split()
        if len(a) != len(b):
            return False
        d1 = {}
        d2 = {}
        for i in zip(a,b):
            d1[i[0]]=i[1]
            d2[i[1]]=i[0]
        return len(d1) == len(d2)

File: algorithms/290.Word_Pattern/Word_Pattern.java
<<<<<<< HEAD
public boolean wordPattern(String pattern, String str) {
    if (pattern == null || str == null) {
        return false;
    }
    char[] patterns = pattern.toCharArray();
    String[] strs = str.split(" ");
    if (patterns.length != strs.length) {
        return false;
    }
    Map<Character, String> map = new HashMap<Character, String>();
    for (int i=0; i<patterns.length; i++) {
        if (map.containsKey(patterns[i])) {
            if (!map.get(patterns[i]).equals(strs[i])) {
                return false;
            }
        } else if (map.containsValue(strs[i])) {
            return false;
        }
        map.put(patterns[i], strs[i]);
    }
    return true;
=======
public boolean wordPattern(String pattern, String str) {
    if (pattern == null || str == null) {
        return false;
    }
    char[] patterns = pattern.toCharArray();
    String[] strs = str.split(" ");
    if (patterns.length != strs.length) {
        return false;
    }
    Map<Character, String> map = new HashMap<Character, String>();
    for (int i=0; i<patterns.length; i++) {
        if (map.containsKey(patterns[i])) {
            if (!map.get(patterns[i]).equals(strs[i])) {
                return false;
            }
        } else if (map.containsValue(strs[i])) {
            return false;
        }
        map.put(patterns[i], strs[i]);
    }
    return true;
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
}